src/Storages/StorageS3.cpp

```diff
@@ -327,6 +344,156 @@ class StorageS3Sink : public SinkToStorage
+            {
+                for (const auto & column_sub_chunk : column_sub_chunks)
+                {
+                    sub_chunks.emplace_back(Columns(), column_sub_chunk->size());
+                }
+            }
+            for (size_t sub_chunk_index = 0; sub_chunk_index < column_sub_chunks.size(); ++sub_chunk_index)
+            {
+                sub_chunks[sub_chunk_index].addColumn(std::move(column_sub_chunks[sub_chunk_index]));
+            }
+        }
+
+        for (const auto & [partition_id, sub_chunk_index] : sub_chunks_indices)
+        {
+            getSinkForPartition(partition_id)->consume(std::move(sub_chunks[sub_chunk_index]));
+        }
+    }
+
+    void onFinish() override
+    {
+        for (auto & [partition_id, sink] : sinks)
+        {
+            sink->onFinish();
+        }
+    }
+
+private:
+    using SinkPtr = std::shared_ptr<StorageS3Sink>;
+
+    const String format;
+    const Block sample_block;
+    ContextPtr context;
+    const CompressionMethod compression_method;
+    std::shared_ptr<Aws::S3::S3Client> client;
+    const String bucket;
+    const String key;
+    size_t min_upload_part_size;
+    size_t max_single_part_upload_size;
+
+    ExpressionActionsPtr partition_by_expr;
+    String partition_by_column_name;
+
+    std::unordered_map<String, SinkPtr> sinks;
+
+    static String replaceWildcards(const String & haystack, const String & partition_id)
+    {
+        return boost::replace_all_copy(haystack, PARTITION_ID_WILDCARD, partition_id);
+    }
+
+    SinkPtr getSinkForPartition(const String & partition_id)
+    {
+        auto it = sinks.find(partition_id);
+        if (it == sinks.end())
+        {
+            std::tie(it, std::ignore) = sinks.emplace(partition_id, std::make_shared<StorageS3Sink>(
+                format,
+                sample_block,
+                context,
+                compression_method,
+                client,
+                replaceWildcards(bucket, partition_id),
+                replaceWildcards(key, partition_id),
+                min_upload_part_size,
+                max_single_part_upload_size
+            ));
+        }
+
+        return it->second;
+    }
+
+    static void validatePartitionKey(const StringRef & str)
+    {
+        for (const char * i = str.data; i != str.data + str.size; ++i)
+        {
+            if (*i < 0x20 || *i == '{' || *i == '}' || *i == '*' || *i == '?')
+            {
+                /// Need to convert to UInt32 because UInt8 can't be passed to format due to "mixing character types is disallowed".
+                UInt32 invalid_char_byte = static_cast<UInt32>(static_cast<UInt8>(*i));
+                throw DB::Exception(ErrorCodes::CANNOT_PARSE_TEXT,
+                        "Illegal character '\\x{0:02x}' in partition key starting with '{}'.",
+                        invalid_char_byte,
+                        StringRef(str.data, i - str.data));
```

Comment:
Strange error `fmt::v7::format_error: cannot switch from manual to automatic argument indexing.`, maybe `StringRef` not supported here
```suggestion
                throw DB::Exception(ErrorCodes::CANNOT_PARSE_TEXT,
                        "Illegal character '\\x{0:02x}' in partition key",
                        invalid_char_byte);
```

Reply:
This is not because of `StringRef`.

Reply:
Actually, new code does not work at all.